### ISSUE TITLE
Session affinity for same service and different ports is forwarded to different endpoints for each destination port

### ISSUE THREAD

**Issue by jcortejoso** [kubernetes/kubernetes#103000]
#### What happened:
Having a service configured with `sessionAffinity: ClientIP`, requests from same client/pod are forwarded to different destination service pod endpoints for each of the service ports.

#### What you expected to happen:
The requests from the same pod/clientIP are forwarded to the same destination pod for all the ports in the service. Quoting the [docs](https://kubernetes.io/docs/concepts/services-networking/service/):

> If you want to make sure that connections from a particular client are passed to the same Pod each time, you can select the session affinity based on the client's IP addresses by setting service.spec.sessionAffinity to "ClientIP" (the default is "None").

#### How to reproduce it (as minimally and precisely as possible):
Create a service with `sessionAffinity: ClientIP` and multiple `spec.ports`. Make requests from a client to the different ports in the service and check if the endpoint for each of the ports is the same pod.

#### Anything else we need to know?:
This is the more meaningful part of my tests:
- Service definition:
```
apiVersion: v1
kind: Service
metadata:
  name: geth-rpc
spec:
  clusterIP: 10.11.250.79
  ports:
  - name: rpc
    port: 8545
    protocol: TCP
    targetPort: rpc
  - name: ws
    port: 8546
    protocol: TCP
    targetPort: ws
  selector:
   ....
  sessionAffinity: ClientIP
  sessionAffinityConfig:
    clientIP:
      timeoutSeconds: 10800
```

Service has 4 endpoints, each of them has a different account that can be returned by the api:
```
root@test-websocket:/# wscat -c ws://geth-rpc:8546
Connected (press CTRL+C to quit)
> {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}
< {"jsonrpc":"2.0","id":64,"result":["0xfbc54f636d7baec867d0a76318d7972005c38f60"]}

> {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}
< {"jsonrpc":"2.0","id":64,"result":["0xfbc54f636d7baec867d0a76318d7972005c38f60"]}

> {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}
< {"jsonrpc":"2.0","id":64,"result":["0xfbc54f636d7baec867d0a76318d7972005c38f60"]}

> {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}
< {"jsonrpc":"2.0","id":64,"result":["0xfbc54f636d7baec867d0a76318d7972005c38f60"]}

> {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}
< {"jsonrpc":"2.0","id":64,"result":["0xfbc54f636d7baec867d0a76318d7972005c38f60"]}
...
Disconnected (code: 1006, reason: "")

root@test-websocket:/# curl -X POST --data '{"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}' http://geth-rpc:8545 -H 'Content-type: application/json'
{"jsonrpc":"2.0","id":64,"result":["0x6b1dfbae1f8c5d183f3e60947ec141c74082b918"]}
root@test-websocket:/# curl -X POST --data '{"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}' http://geth-rpc:8545 -H 'Content-type: application/json'
{"jsonrpc":"2.0","id":64,"result":["0x6b1dfbae1f8c5d183f3e60947ec141c74082b918"]}
root@test-websocket:/# curl -X POST --data '{"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}' http://geth-rpc:8545 -H 'Content-type: application/json'
{"jsonrpc":"2.0","id":64,"result":["0x6b1dfbae1f8c5d183f3e60947ec141c74082b918"]}
root@test-websocket:/# curl -X POST --data '{"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}' http://geth-rpc:8545 -H 'Content-type: application/json'
{"jsonrpc":"2.0","id":64,"result":["0x6b1dfbae1f8c5d183f3e60947ec141c74082b918"]}
root@test-websocket:/# curl -X POST --data '{"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":64}' http://geth-rpc:8545 -H 'Content-type: application/json'
{"jsonrpc":"2.0","id":64,"result":["0x6b1dfbae1f8c5d183f3e60947ec141c74082b918"]}
...

```

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:11:29Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.17-gke.4900", GitCommit:"2812f9fb0003709fc44fc34166701b377020f1c9", GitTreeState:"clean", BuildDate:"2021-03-19T09:19:27Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: GKE default setup. Istio disabled.
- OS (e.g: `cat /etc/os-release`):
```
BUILD_ID=12871.1245.7
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=23851459ab405049e5ee00aa54b3c3ef32abad6d
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=81
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=81
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
ID=cos
```
- Kernel (e.g. `uname -a`): `Linux gke-<instance_name> 4.19.167+ #1 SMP Tue Feb 2 07:59:55 PST 2021 x86_64 Intel(R) Xeon(R) CPU @ 2.20GHz GenuineIntel GNU/Linux`
- Install tools: GKE
- Network plugin and version (if this is a network-related bug): GPC network
- Others:

**Comment by uablrek** [comment]
What proxy-mode are you using, ipvs or iptables?

For ipvs the different service ports becomes different "targets" and affinity is per target (I think). It might be hard to fix, but may be simpler for proxy-mode=iptables.

**Comment by jcortejoso** [comment]
Under my understanding it is using iptables mode (and GKE does not support directly ipvs mode).:

```
$ kubectl describe pod kube-proxy-gke-mainnet-e2-standard-8-d4f8f84d-44cf -n kube-system
Name:                 kube-proxy-gke-e2-standard-8-d4f8f84d-44cf
Namespace:            kube-system
Priority:             2000001000
...
Controlled By:  Node/gke-mainnet-e2-standard-8-d4f8f84d-44cf
Containers:
  kube-proxy:
    Image:         gke.gcr.io/kube-proxy-amd64:v1.17.17-gke.4900
...
    Command:
      /bin/sh
      -c
      exec kube-proxy --master=https://$MASTER_IP --kubeconfig=/var/lib/kube-proxy/kubeconfig --cluster-cidr=10.8.0.0/14 --oom-score-adj=-998 --v=2 --feature-gates=DynamicKubeletConfig=false,RotateKubeletServerCertificate=true --iptables-sync-period=1m --iptables-min-sync-period=10s --ipvs-sync-period=1m --ipvs-min-sync-period=10s 1>>/var/log/kube-proxy.log 2>&1
    State:          Running
...
    HostPathType:
  iptableslock:
    Type:          HostPath (bare host directory volume)
    Path:          /run/xtables.lock
    HostPathType:  FileOrCreate
...
```

Another consideration is that if I ssh to the node that is running the `test-websocket` pod, I cannot see or identify the iptable rules corresponding to the service `geth-rpc`:

```
-A KUBE-EXTERNAL-SERVICES -p udp -m comment --comment "geth/fullnodes-2-udp:ethereum has no endpoints" -m addrtype --dst-type LOCAL -m udp --dport 32605 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-EXTERNAL-SERVICES -p udp -m comment --comment "geth/fullnodes-0-udp:ethereum has no endpoints" -m addrtype --dst-type LOCAL -m udp --dport 32437 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-EXTERNAL-SERVICES -p udp -m comment --comment "geth/fullnodes-1-udp:ethereum has no endpoints" -m addrtype --dst-type LOCAL -m udp --dport 31941 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-EXTERNAL-SERVICES -p udp -m comment --comment "geth/fullnodes-3-udp:ethereum has no endpoints" -m addrtype --dst-type LOCAL -m udp --dport 30227 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-FIREWALL -m comment --comment "kubernetes firewall for dropping marked packets" -m mark --mark 0x8000/0x8000 -j DROP
-A KUBE-FIREWALL ! -s 127.0.0.0/8 -d 127.0.0.0/8 -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
-A KUBE-FORWARD -m conntrack --ctstate INVALID -j DROP
-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x4000/0x4000 -j ACCEPT
-A KUBE-FORWARD -s 10.8.0.0/14 -m comment --comment "kubernetes forwarding conntrack pod source rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A KUBE-FORWARD -d 10.8.0.0/14 -m comment --comment "kubernetes forwarding conntrack pod destination rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A KUBE-SERVICES -d 10.11.253.225/32 -p udp -m comment --comment "geth/fullnodes-2-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 35.247.83.16/32 -p udp -m comment --comment "geth/fullnodes-2-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 10.11.246.152/32 -p udp -m comment --comment "geth/fullnodes-0-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 35.230.28.149/32 -p udp -m comment --comment "geth/fullnodes-0-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 10.11.246.224/32 -p udp -m comment --comment "geth/fullnodes-1-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 35.230.112.14/32 -p udp -m comment --comment "geth/fullnodes-1-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 10.11.244.198/32 -p udp -m comment --comment "geth/fullnodes-3-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 34.82.120.112/32 -p udp -m comment --comment "geth/fullnodes-3-udp:ethereum has no endpoints" -m udp --dport 30303 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-SERVICES -d 10.11.242.20/32 -p tcp -m comment --comment "default/block-producer-beta:tcp-p2p has no endpoints" -m tcp --dport 10101 -j REJECT --reject-with icmp-port-unreachable
```

**Comment by thockin** [comment]
This is a legit bug in iptable.  When we do a "recent" check, we check per SEP (service endpoint - aka port) not per service.

The question is whether we can actually fix it or not.   In particular, it is possible to have a backend pod which offers some but not all service ports (via `targetPort=<string>`) and fixing this would break clients.  It seems EXCEEDIGLY unlikely to be a real problem, but it could be.  Also see comments above about IPVS.

If IPVS could support it, I could see maybe fixing this and adding a different option for per-client-per-port.

Alternatively, we just fix the docs.

Can you tell us more about the use-case?

/triage accepted

**Comment by aojea** [comment]
I think that we can justify one or the other option, to me this is a problem of how do we want to interpret a Service in Kubernetes

Per example, assume a service with ClusterIP and Port1 and Port2, this can be translated to:
1. ClusterIP:Port1-> Pod1:PortX, Pod2:PortX and ClusterIP:Port2 -> Pod1:PortY, Pod2:PortY
2. ClusterIP:[Port1,Port2] --> Pod1:[PortX,PortY] and Pod2:[PortX,PortY]

Most loadbalancers, and kube-proxy in both flavors, interprets (right now) the Service as 1. however, documentations is not reflecting that as this issue explains. The endpoints are different and independent.

In 2. it seems we are implying that there is a relation between the ports and the back end

I'm in favour of fixing the docs because it reflects current behaviour and the most common setup for networking loadbalancers, although I consider 2. a valid interpretation too

**Comment by uablrek** [comment]
I am unsure if my https://github.com/kubernetes/kubernetes/issues/103000#issuecomment-864365374 is valid. Reading;

https://docs.huihoo.com/hpc-cluster/linux-virtual-server/HOWTO/LVS-HOWTO.persistent_connection.html

I can't see immediately if ipvs supports the "full service redirect", requested in this issue, by default. It is a TL;DR page and I think the best is to test it in K8s. I will do that, but not today since it's midsummers eve and all Swedes perform the [little frogs](https://en.wikipedia.org/wiki/Sm%C3%A5_grodorna) ritual in the light of the midnight sun (well, at least we don't work).

**Comment by uablrek** [comment]
Seem like proxy-mode=ipvs _have_ affinity for _all_ ports in the service. So the function exist in proxy-mode=ipvs but not in proxy-mode=iptables. I have tested with;
```yaml
apiVersion: v1
kind: Service
metadata:
  name: mserver-affinity
spec:
  sessionAffinity: ClientIP
  sessionAffinityConfig:
    clientIP:
      timeoutSeconds: 120
  ipFamilyPolicy: PreferDualStack
  selector:
    app: mserver
  ports:
  - port: 5001
    name: mconnect
  - port: 5003
    name: ctraffic
  type: LoadBalancer
```
When a client makes a connect to 5001, subsequent connects to 5001 _and_ 5003 goes to the same server.

Note that `type: LoadBalancer` does not permit multi protocols, so test the combo TCP/UDP is not possible.

**Comment by uablrek** [comment]
I made an unplesant discovery; I use ECMP to distribute traffic between nodes and the `ipvs` affinity state ("template") is per node. So subsequent calls will go to another server if ECMP directs it to another node.

It is as it must be since there is no coordination between `ipvs` instances on different nodes.

**Comment by aojea** [comment]
> I made an unplesant discovery; I use ECMP to distribute traffic between nodes and the `ipvs` affinity state ("template") is per node. So subsequent calls will go to another server if ECMP directs it to another node.
> 
> It is as it must be since there is no coordination between `ipvs` instances on different nodes.

heh, it should be the same for iptables, I don't think that iptables keeps state between nodes :upside_down_face:

**Comment by jcortejoso** [comment]
> Can you tell us more about the use-case?

The use-case I was testing here is a service that subscribes to a websocket to get "new blocks" events, and when it gets that event it tries to fetch the block from the node using an rpc call. So if the rpc and websocket and rpc ports are served by different ports, the server responding to the requests can be different, and in this case (ethereum nodes) there could be some delays between when different nodes syncs the new blocks, so I was getting some 404 in the rpc calls.
Using an nginx or other proxy solution as a sidecar can be used as a workaround for these scenarios, but reading the documentation I'd expect all the ports in the same service to be attached to the same endpoint if sessionAffinity is enabled.

**Comment by aojea** [comment]
> the server responding to the requests can be different, and in this case (ethereum nodes) there could be some delays between when different nodes syncs the new blocks, so I was getting some 404 in the rpc calls.

does this means that, in theory, with current behavior, it should work but it doesn't because the nodes backing the service have some delay synchronizing their state?

> but reading the documentation I'd expect all the ports in the same service to be attached to the same endpoint if sessionAffinity is enabled.

indeed, I think that the bug is acknowledged, the debate now is if it is a bug in the docs or a bug in the code :sweat_smile: 

> ; I use ECMP to distribute traffic between nodes and the `ipvs` affinity state ("template") is per node.

But based on Lars finding, it seems that the affinity is always per node, if the external client "enters" the cluster using a different node you are always going to hit a random backend.
That reinforces my opinion that this is a documentation bug, and the session affinity should be by ClusterIP:Port not by ClusterIP only, since it is relatively easy to guarantee that the same TCP tuple is redirected always to the same node, hence the affinity for the same tuple will work

**Comment by jcortejoso** [comment]
> does this means that, in theory, with current behavior, it should work but it doesn't because the nodes backing the service have some delay synchronizing their state?

Yes, exactly that is the situation.

> indeed, I think that the bug is acknowledged, the debate now is if it is a bug in the docs or a bug in the code 😅

Right. In my personal opinion seems ok to me that kubernetes services do not support session affinity on different target ports, so it would look to me a right solution updating the documentation with a clarification over the matter. In this case the consumption of the service is internal so I have not faced the problems of the affinity is in per node basis, but yeah I agree on fixing the documentation would be a good solution.

**Comment by thockin** [comment]
Interesting.  Does IPVS have a way to make that independent per-port
(as a choice) or is that just fixed in the IPVS mechanism?

Some LB mechanisms can do affinity and some can't, which isn't
surprising.  If I understand, ECMP + IPVS approximates the iptables
per-port affinity.  But a more stateful routing layer wouId
approximate the IPVS per-service affinity.  I don't think we can
mandate better-than-ECMP, but we shouldn't dumb it down if we can
avoid it.

So, if we "fix" iptables to be per-service, it's a notable behavioral
change.  We could *add* options.  E.g. Google's LBs have options for
client IP, client IP + protocol, client IP + protocol + port.  We
could maybe do something like:

```
const (
    // ServiceAffinityClientAddr considers the client's IP address for
session affinity.
    ServiceAffinityClientAddr ServiceAffinity = "ClientAddr"

    // ServiceAffinityClientAddrProto considers the client's IP
address and protocol for session affinity.
    ServiceAffinityClientAddrProto ServiceAffinity = "ClientAddrProto"

    // ServiceAffinityClientAddrProtoPort considers the client's IP
address, protocol, and port number for session affinity.
    ServiceAffinityClientAddrProtoPort ServiceAffinity = "ClientAddrProtoPort"

    // ServiceAffinityClientIP is ambiguous.  Different network
configurations (e.g. kube-proxy setting) can produce different
behaviors.
    // This option is deprecated.  Callers should prefer the above
values which are more explicit.
    ServiceAffinityDeprecatedClientIP ServiceAffinity = "ClientIP"

    // ServiceAffinityNone - no session affinity.
    ServiceAffinityNone ServiceAffinity = "None"
)
```

To do this, we'd need to do some homework on implementations.  Hence
my questions about IPVS above.

**Comment by uablrek** [comment]
> Does IPVS have a way to make that independent per-port (as a choice) or is that just fixed in the IPVS mechanism?

From reading a bit more in 
https://docs.huihoo.com/hpc-cluster/linux-virtual-server/HOWTO/LVS-HOWTO.persistent_connection.html

in section 21.1; 

> client is recognised by its IP (CIP) or in recent versions of ip_vs, by CIP:dst_port (i.e. by the CIP and the port being forwarded by the LVS)

it looks like the default is to _not_ direct all ports from a client to the same server, the opposite from what I observed. I made 100 connects to port 5001 verifying that they go to the same server and then _one_ connect to 5003. I made that 10 times in a row from different client IP's and the connect to 5003 always got to the same server as the connects to 5001. I also tested with proxy-mode=iptables and the call to 5003 rarely came to the same server (as expected).

To get all ports to the same server in ipvs you should specify VIP:0 if I get the docs right. K8s does not do that.

**Comment by uablrek** [comment]
An application must handle the case where connects happen to come to the same server in any case so I can't see any problem there. The problem might be uneven load-balancing, but I think that is a minor problem.

**Comment by thockin** [comment]
If we ever implement all-ports, it seems VIP:0 + persistence is the
only option for IPVS.  But I think I agree that, as a special case,
that is OK.

So If I interpret correctly, IPVS is *supposed* to be affinity
per-port (though it's not clear it actually does that?).  As such, we
should just fix our docs.  Is that the right take-0away, Lars?

Thanks for investigating!

**Comment by uablrek** [comment]
Yes, as far as I can understand the ipvs docs.

**Comment by srampal** [comment]
@jcortejoso @thockin  Imo this is a very valid issue and should not just be a documentation change imo .. (unless I am missing something about the exact scenario). 

Was the issue seen when the client was accessing the service from outside the cluster ? or was the client a pod in the same cluster ? I thought it was the latter and if so,  this is not just a documentation issue per my understanding since all traffic from such a client should go to the same backend pod for both ports of the clusterIP service as you also felt. This seems to be a common use case since a client may be running an app that depends on multiple ports which should be on the same actual backend pod.  Imo, whatever is implementing the service load balancing whether it is iptables or IPVS or vendor CNI plugin switch), should ensure all traffic from a single intra-cluster client towards all ports of a clusterIP should end up at the same backend pod when sessionAffinity is set to clientIP. Maybe another value for sessionAffinity should be defined in order to allow traffic to go to different backend pods even if multiple destination ports are for the same clusterIP service.

Let me know in case I am missing something :-).

**Comment by thockin** [comment]
Revisiting this.  My summary:

* kube-proxy iptables does affinity per-client, per-port.  No cross-port affinity.
* kube-proxy IPVS is documented to do affinity per-client, per-port, but it's not clear if reality matches the docs.  Need clarity.
* Do we know what other implementations do for this?  @howardjohn?  @tgraf?
* The feature request is to enable cross-port affinity.
* To really make this work, intermediate LBs (e.g. clouds) need to ALSO do whole-IP affinity.  CCM configures GCE's L4 LB for this today.  I don't know the others.
* I'd be OK, I think, to add one or more options, but I don't think we can change the meaning of the existing value. Whether we can do this depends on implementations' ability to provide the right semantics.  I know iptables can.  I think IPVS can only do it for whole-IPs (which means ports would be exposed that were not intended) and it's not clear if port remapping (IPVS masquerade) will work.  That alone may kill the idea

**Comment by uablrek** [comment]
Add;
* If ecmp on between nodes is used service affinity can't be guaranteed

For IPv4 it "should work" with the default (L3);
```
fib_multipath_hash_policy - INTEGER
	Controls which hash policy to use for multipath routes. Only valid
	for kernels built with CONFIG_IP_ROUTE_MULTIPATH enabled.

	Default: 0 (Layer 3)

	Possible values:

	- 0 - Layer 3
	- 1 - Layer 4
	- 2 - Layer 3 or inner Layer 3 if present
```
For IPv6 it becomes more complex since the "flow label" is included;
```
fib_multipath_hash_policy - INTEGER
        Controls which hash policy to use for multipath routes.

        Default: 0 (Layer 3)

        Possible values:

        - 0 - Layer 3 (source and destination addresses plus flow label)
        - 1 - Layer 4 (standard 5-tuple)
        - 2 - Layer 3 or inner Layer 3 if present
```

**Comment by uablrek** [comment]
Next step for `ipvs` should be to check the kernel source. I have not done that for like >8years, perhaps @andrewsykim can take a look at it?

My experiment described in https://github.com/kubernetes/kubernetes/issues/103000#issuecomment-870746060 gives that `ipvs` _has_ service affinity, but the documentations says that is doesn't (but not very clear imho).

**Comment by uablrek** [comment]
Linux kernel `auto_flowlabels`:
```
auto_flowlabels - INTEGER
        Automatically generate flow labels based on a flow hash of the
        packet. This allows intermediate devices, such as routers, to
        identify packet flows for mechanisms like Equal Cost Multipath
        Routing (see RFC 6438).

        =  ===========================================================
        0  automatic flow labels are completely disabled
        1  automatic flow labels are enabled by default, they can be
           disabled on a per socket basis using the IPV6_AUTOFLOWLABEL
           socket option
        2  automatic flow labels are allowed, they may be enabled on a
           per socket basis using the IPV6_AUTOFLOWLABEL socket option
        3  automatic flow labels are enabled and enforced, they cannot
           be disabled by the socket option
        =  ===========================================================

        Default: 1
```

**Comment by aojea** [comment]
> * To really make this work, intermediate LBs (e.g. clouds) need to ALSO do whole-IP affinity.  CCM configures GCE's L4 LB for this today.  I don't know the others.

That is a big assumption hard to generalize, I think that this is much simpler to solve one layer above of Service, using Ingress or Gateway API.

**Comment by uablrek** [comment]
> I think that this is much simpler to solve one layer above of Service, using Ingress or Gateway API.

Would that not require all ports to be directed to the same ingress controller? If so, that would return the requirement to service level. It can easily be accomplished though with a daemonset and `externalTrafficPolicy: Local` (except for the ecmp problem). Perhaps that is enough? If apps that need full service affinity can live with that constraint the problem is solved, or at least a work-around exist.

**Comment by howardjohn** [comment]
> Do we know what other implementations do for this?

Istio does it per-port

**Comment by aojea** [comment]
> Would that not require all ports to be directed to the same ingress controller? If so, that would return the requirement to service level

yeah, you are right, I was assuming only one ingress instance :facepalm: 

I'm trying to understand what applications will need this feature, so we don't fall in the trap of "use the network to workaround the app limitations", where things that always end bad.

I don't think that FTP is a good example, it will require always a singleton ...

**Comment by Nish0808** [comment]
I am also facing this issue which is mentioned here:
sessionAffinity based on ClientIP, is not maintained for a new UDP request followed by TCP request from same client #104659
Do we have any workaround  at Kubernetes level ?

**Comment by aojea** [comment]
> I am also facing this issue which is mentioned here:
> sessionAffinity based on ClientIP, is not maintained for a new UDP request followed by TCP request from same client #104659
> Do we have any workaround at Kubernetes level ?

can you explain the use case better? 
what type of application and deployment are you using that requires UDP and TCP session affinity?

**Comment by Nish0808** [comment]
@aojea Updated my ticket with use case.

**Comment by khenidak** [comment]
i don't think it is possible to maintain affinity across all ports for a service. I can tell with high degree of certainty (but not with absolute confidence) that both iptables/ipvs can not do it. The same is true (with absolute confidence) for typical LBs affinity is per rule (which includes target port), you can not have affinity across two distinct rules (e.g. one rule for each port).
